format/FlightSql.proto

```diff
@@ -1778,6 +1794,47 @@ message CommandPreparedStatementUpdate {
   bytes prepared_statement_handle = 1;
 }
 
+/*
+ * Represents a bulk ingestion request. Used in the command member of FlightDescriptor
+ * for the the RPC call DoPut to cause the server load the contents of the stream's
+ * FlightData into the target destination.
+ */
+message CommandStatementIngest {
+  option (experimental) = true;
+
+  // Describes the behavior for loading bulk data.
+  enum IngestMode {
+    // Ingestion behavior unspecified.
+    INGEST_MODE_UNSPECIFIED = 0;
+    // Create the target table. Fail if the target table already exists.
+    INGEST_MODE_CREATE = 1;
+    // Append to an existing target table. Fail if the target table does not exist.
+    INGEST_MODE_APPEND = 2;
+    // Drop the target table if it exists. Then follow INGEST_MODE_CREATE behavior.
+    INGEST_MODE_REPLACE = 3;
+    // Create the target table if it does not exist. Then follow INGEST_MODE_APPEND behavior.
+    INGEST_MODE_CREATE_APPEND = 4;
+  }
+
+  // The ingestion behavior.
+  IngestMode mode = 1;
+  // The table to load data into.
+  string table = 2;
+  // The db_schema of the destination table to load data into. If unset, a backend-specific default may be used.
+  optional string schema = 3;
+  // The catalog of the destination table to load data into. If unset, a backend-specific default may be used.
+  optional string catalog = 4;
+  // Use a temporary table.
+  optional bool temporary = 5;
```

Comment:
Can you explain why it would be useful? Isn't this more of a server-specific implementation detail?

Reply:
This is aligned with the current functionality of ABDC. There is some helpful context on the meaning and usefulness of this parameter in the [original issue](https://github.com/apache/arrow-adbc/issues/1018).

Reply:
Ah! Is this supposed to mean that _if a new table is created for ingestion_, then the table is made temporary?

The comment is very confusing. I thought it meant that the ingestion was first done into a temporary table, then committed into the actual result table.

Reply:
The comment now reads: `Store ingested data in a temporary table`

How does that sound? I wanted to avoid specifically stating that a temporary table would be **created** because the specific choice of `mode` might contradict this behavior.

Reply:
Then what does it do if I pass INGEST_MODE_APPEND?

Reply:
You can append data to an existing temporary table.

Reply:
Ok, but what does it do if the table is not temporary?

Seriously, please spell out the semantics explicit in the comment. This is not helpful.

Reply:
I took some time to map out how this could work in the various ingestion scenarios and ran into some trouble with certain edge cases. After reviewing how several DB engines I'm familiar with handle these scenarios, I can see the benefit of making an opinionated choice regarding the semantics.

Based on my observations, there are at least **two** qualities of temporary tables that most clients would expect:
1. They are dropped automatically by the DB at the end of the session. The scope of the session could be a connection, transaction, etc. The specific scope seems to be backend-specific, and sometimes configurable.
2. Another quality that seems common to implementations I've seen is that you treat temp tables _exactly the same_ as a "regular" table after they've been created, for the duration of the session. This means that the only time a client would specify that a table is temporary is during creation (i.e. `CREATE` vs `CREATE TEMP`). For this reason I take back my statement implying that `temporary` is unrelated to table creation. To the contrary it seems that table creation is the only place where `temporary` should have an effect.

This is based on DBs I'm familiar with, so I would appreciate your perspectives on whether these expectations are broadly applicable.

Putting these observations together, I would propose the following semantics when used with `temporary`:
- `INGEST_MODE_CREATE`: Create a temp table. Fail if it already exists. Should create even if a non-temp table with the same name already exists.
- `INGEST_MODE_APPEND`: Ignore the temporary argument. Append to the table specified. The backend can tell whether it's temporary or not. (I'm ok with returning an error here instead if there's a good reason to do so)
- `INGEST_MODE_REPLACE`: Drop the table if it exists. Then follow INGEST_MODE_CREATE behavior. Fail if the target is not a temp table (might help avoid unexpected data loss).
- `INGEST_MODE_CREATE_APPEND`: Create a temp table if it does not exist. Then follow INGEST_MODE_APPEND behavior.

Thoughts on these conclusions?

Reply:
Temporary tables appear to generally be treated as if they're under a different namespace (so in some sense the flag is redundant with the catalog/db_schema arguments, though it saves you from knowing exactly what the database calls it and whether it's a schema or a catalog). So I `APPEND` should fail if no temporary table exists by that name, and that `REPLACE` would ignore a non-temporary table that otherwise has that name (and thereby fail if there is no temporary table with that name).

Reply:
I agree with this. There is a remaining edge case with `APPEND` that I think is worth clarifying:
1. Create and write into a temp table `foo` (mode=`CREATE`, table=`foo`, temporary=`true`)
2a. Append to `foo` without specifying temporary in the request (mode=`APPEND`, table=`foo`), or...
2b. Append to `foo` with temporary `false` (mode=`APPEND`, table=`foo`, temporary=`false`)

Even if there is an existing non-temp table `foo`, I think most clients would expect the creation in step 1 to succeed, shadowing the original table.

Subsequent appends that omit the `temporary` arg (case 2a) would continue to exhibit this shadowing behavior, targeting the temp table as long as it exists in the session. This behavior would be identical in this case to setting temporary=`true`, except when it is explicitly `true` an additional check is performed to ensure the table is temporary, failing if it is not.

If subsequent appends explicitly set temporary=`false` (case 2b), then the data is appended to the non-temp table with that name. If a non-temp table with that name does not exist, the request fails.

A concrete example (with equivalent pseudo-SQL):
```proto
// Create temp and non-temp
CommandStatementIngest(mode=CREATE, table=foo) -> CREATE TABLE foo
CommandStatementIngest(mode=CREATE, table=foo, temporary=true) -> CREATE TEMP TABLE foo

// Only create non-temp
CommandStatementIngest(mode=CREATE, table=bar) -> CREATE TABLE bar

// Only create temp
CommandStatementIngest(mode=CREATE, table=baz, temporary=true) -> CREATE TEMP TABLE baz

// Temp unspecified
CommandStatementIngest(mode=APPEND, table=foo) -> INSERT INTO temp.foo // shadowing public.foo
CommandStatementIngest(mode=APPEND, table=bar) -> INSERT INTO public.bar
CommandStatementIngest(mode=APPEND, table=baz) -> INSERT INTO temp.baz

// Temp true
CommandStatementIngest(mode=APPEND, table=foo, temporary=true) -> INSERT INTO temp.foo
CommandStatementIngest(mode=APPEND, table=bar, temporary=true) -> ERROR // not temporary
CommandStatementIngest(mode=APPEND, table=baz, temporary=true) -> INSERT INTO temp.baz

// Temp false
CommandStatementIngest(mode=APPEND, table=foo, temporary=false) -> INSERT INTO public.foo
CommandStatementIngest(mode=APPEND, table=bar, temporary=false) -> INSERT INTO public.bar
CommandStatementIngest(mode=APPEND, table=baz, temporary=false) -> ERROR // only temporary
```

I think this behavior would be the most intuitive based on the arguments we expose. The last case where temporary=`false` is currently achieved in most systems by explicitly specifying the namespace of the table (e.g. `public.foo` instead of `foo`). So it is worth pointing out that we would be adding a feature that isn't necessarily common (explicitly unshadow a non-temp table without knowing the original schema name), but it may be useful and likely wouldn't be difficult to implement even if not directly supported by the underlying database.

Reply:
I updated the proto comment to capture these semantics.

Reply:
> So it is worth pointing out that we would be adding a feature that isn't necessarily common (explicitly unshadow a non-temp table without knowing the original schema name)

The ADBC drivers supporting this feature do this too. Basically, `temporary=true` is equivalent to setting `schema=pg_temp`, and conversely `temporary=false` is equivalent to setting `schema=CURRENT_SCHEMA`.